lib/atoi/strtou_noneg.h

```diff
@@ -0,0 +1,47 @@
+// SPDX-FileCopyrightText: 2023, Alejandro Colomar <alx@kernel.org>
+// SPDX-License-Identifier: BSD-3-Clause
+
+
+#ifndef SHADOW_INCLUDE_LIB_ATOI_STRTOU_NONEG_H_
+#define SHADOW_INCLUDE_LIB_ATOI_STRTOU_NONEG_H_
+
+
+#include <config.h>
+
+#include <errno.h>
+#include <stdlib.h>
+
+#include "attr.h"
+
+
+ATTR_STRING(1) ATTR_ACCESS(write_only, 2)
+inline unsigned long strtoul_noneg(const char *nptr,
+    char **restrict endptr, int base);
+ATTR_STRING(1) ATTR_ACCESS(write_only, 2)
+inline unsigned long long strtoull_noneg(const char *nptr,
+    char **restrict endptr, int base);
+
+
+inline unsigned long
+strtoul_noneg(const char *nptr, char **restrict endptr, int base)
+{
+	if (strtol(nptr, endptr, base) < 0) {
```

Comment:
I know you are just replacing `strtoul` calls by this wrapper in the code, but shouldn't we check for overflows and underflows?

Reply:
strtoul(3) already checks for overflow.  With that call, I add a check for underflow.

Or do you mean something else?  Can you please clarify?

Reply:
strtoul() checks for overflows, and in that case it returns a specific value (ULONG_MAX and sets errno to ERANGE), but you aren't checking for those values.

Reply:
Ahh, so you mean in the calls to strtoul_noneg().  All such calls are guarded by `if (errno != 0)`, so they do check for overflow.  strtoul_noneg(), like strtol(3), sets errno to ERANGE on overflow or underflow.

Reply:
The following PR is related: <https://github.com/shadow-maint/shadow/pull/869>

Reply:
:O I see! That check is being done by the caller.